cpu/rpx0xx/periph/timer.c

```diff
@@ -0,0 +1,241 @@
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_rp2040
+ * @ingroup     drivers_periph_timer
+ * @{
+ *
+ * @file
+ * @brief       Timer implementation for the RPX0XX
+ * @details     The RPX0XX has a 64 bit µs timer but timer interrupts match
+ *              on the lower 32 bits.
+ *
+ * @author      Fabian Hüßler <fabian.huessler@ovgu.de>
+ *
+ * @}
+ */
+
+#include <errno.h>
+#include <assert.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+#include "vendor/RP2040.h"
+#include "io_reg.h"
+#include "timex.h"
+#include "periph_conf.h"
+#include "periph/timer.h"
+
+#define DEV(d)          (timer_config[d].dev)
+#define ALARM(d, a)     ((&(DEV(d)->ALARM0)) + (a))
+
+static timer_cb_t _timer_ctx_cb[TIMER_NUMOF];
+static void *_timer_ctx_arg[TIMER_NUMOF];
+static unsigned _timer_flag_periodic[TIMER_NUMOF];
+static unsigned _timer_flag_reset[TIMER_NUMOF];
+
+static inline uint64_t _timer_read_us(tim_t dev)
+{
+    (void)dev;
+    uint32_t time_h, tmp = TIMER->TIMERAWH;
+    uint32_t time_l;
+    do {
+        time_h = tmp;
+        time_l = TIMER->TIMERAWL;
+    } while ((tmp = TIMER->TIMERAWH) != time_h);
+    return (((uint64_t)time_h) << 32) | time_l;
+}
+
+static inline void _timer_reset(tim_t dev)
+{
+    DEV(dev)->TIMELW = 0; /* always write timelw before timehw */
+    DEV(dev)->TIMEHW = 0; /* writes do not get copied to time until timehw is written */
+}
+
+static inline void _timer_enable_periodic(tim_t dev, int channel, uint8_t flags)
+{
+    _timer_flag_periodic[dev] |= (1U << channel);
+    if (flags & TIM_FLAG_RESET_ON_MATCH) {
+        _timer_flag_reset[dev] |= (1U << channel);
+    }
+    else {
+        _timer_flag_reset[dev] &= ~(1U << channel);
+    }
+}
+
+static inline void _timer_disable_periodic(tim_t dev, int channel)
+{
+    _timer_flag_periodic[dev] &= ~(1U << channel);
+}
+
+static inline bool _timer_is_periodic(tim_t dev, int channel)
+{
+    return !!(_timer_flag_periodic[dev] & (1U << channel));
+}
+
+static inline bool _timer_reset_on_match(tim_t dev, int channel)
+{
+    return !!(_timer_flag_reset[dev] & (1U << channel));
+}
+
+static inline void _irq_enable(tim_t dev)
+{
+    for (uint8_t i = 0; i < timer_config[dev].chn; i++) {
+        NVIC_EnableIRQ(timer_config[dev].ch[i].irqn);
+        io_reg_atomic_set(&DEV(dev)->INTE.reg, (1U << i));
+    }
+}
+
+static void _isr(tim_t dev, int channel)
+{
+    /* clear latched interrupt */
+    io_reg_atomic_clear(&DEV(dev)->INTR.reg, 1U << channel);
+
+    if (_timer_ctx_cb[dev]) {
+        _timer_ctx_cb[dev](_timer_ctx_arg[dev], channel);
+    }
+
+    if (_timer_is_periodic(dev, channel)) {
+        if (_timer_reset_on_match(dev, channel)) {
+            _timer_reset(dev);
+        }
+        /* rearm */
+        *ALARM(dev, channel) = *ALARM(dev, channel);
+    }
```

Comment:
Shouldn't this be done first in the ISR?

On most MCUs, the reset on match would be done by the hardware right after the alarm value was hit, so that e.g. a periodic timer with reset on match at 10 ms would result in an IRQ precisely every 10 ms. Since the timer of the RP2040 is not able to do this in hardware, we should emulate the reset as closely as possible to the hardware behavior. Doing the reset right at the beginning of the ISR should increase the odds that the reset is happening before the timer increments after the match.

Alternatively, one could do something like this:
1. Busy wait until the timer ticks
2. Set the timer to the value of microseconds that have passed since the alarm to compensate for the ISR duration

But this could be a bit tricky to implement correctly.